torch/testing/_internal/common_methods_invocations.py

```diff
@@ -345,6 +345,94 @@ def __init__(self,
         #   outside a function's domain.
         self._domain_eps = 1e-5
 
+
+def sample_inputs_binary(op_info, device, dtype, requires_grad):
+    low, high = op_info.domain
+    low = low if low is None else low + op_info._domain_eps
+    high = high if high is None else high - op_info._domain_eps
+
+    return (
+        # empty tensors
+        SampleInput((make_tensor((), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad),
+                     make_tensor((), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad))),
+        # 1d - 1d
+        SampleInput((make_tensor((L), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad),
+                     make_tensor((L), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad))),
+        # 1d - 2d
+        SampleInput((make_tensor((L), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad),
+                     make_tensor((L, L + 5), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad))),
+        # 2d - 1d
+        SampleInput((make_tensor((L, L + 5), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad),
+                     make_tensor((L), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad))),
+        # 2d - 2d
+        SampleInput((make_tensor((L, L + 5), device, dtype,
+                                 low=low, high=high,
+                                 requires_grad=requires_grad),
+                     make_tensor((L, L + 5), device, dtype,
```

Comment:
If we look at the method_tests() for add:

```
('add', (S, S, S), ((S, S, S),), '', (True,)),
('add', (S, S, S), ((S, S),), 'broadcast_rhs', (True,)),
('add', (S, S), ((S, S, S),), 'broadcast_lhs', (True,)),
('add', (S, 1, S), ((M, S),), 'broadcast_all', (True,)),
('add', (), ((),), 'scalar', (True,)),
('add', (S, S, S), ((),), 'scalar_broadcast_rhs', (True,)),
('add', (), ((S, S, S),), 'scalar_broadcast_lhs', (True,)),
('add', (S, S, S), (3.14,), 'constant', (True,)),
('add', (), (3.14,), 'scalar_constant', (True,)),
('add', (S, S, S), (3.14j,), 'complex_scalar_constant', (True,)),
```

I would expect this to have the same. That is (I've re-ordered the above cases):

```
# scalar tensor x Python scalar (if complex. python scalar should be complex, if int, python scalar should be int)
('add', (), (3.14,), 'scalar_constant', (True,)),

# scalar tensor x scalar tensor
('add', (), ((),), 'scalar', (True,)),

# scalar tensor x 3d
('add', (), ((S, S, S),), 'scalar_broadcast_lhs', (True,)),

# 3d scalar tensor
('add', (S, S, S), ((),), 'scalar_broadcast_rhs', (True,)),

# 3d x python scalar (python scalar should be typed as per dtype)
('add', (S, S, S), (3.14j,), 'complex_scalar_constant', (True,)),
('add', (S, S, S), (3.14,), 'constant', (True,)),

# 3d x 2d
('add', (S, S, S), ((S, S),), 'broadcast_rhs', (True,)),

# 2d x 3d
('add', (S, S), ((S, S, S),), 'broadcast_lhs', (True,)),

# 3d x 3d
('add', (S, S, S), ((S, S, S),), '', (True,)),

# 3d x 2d broadcasting
('add', (S, 1, S), ((M, S),), 'broadcast_all', (True,)),
```

One reasonably quick way of creating these cases is to do something like

```
test_cases = (
  ((S, 1, S), (M, S)),
  ...
)

for test_case in test_cases:
  # instantiate SampleInput here
```

Using that pattern would require adding the Python scalar test case separately. Note that we don't support the first argument to an operation not being a tensor currently, so this should probably limit itself to tensor x python scalar and not pursue python scalar x tensor for now.

Does that make sense?

Reply:
yes

Reply:
Used the paradigm in 4ae138e, without python scalars yet